File: src/main/org/nlogo/prim/gui/_load3Dshapes.java
package org.nlogo.prim.gui ;

import org.nlogo.api.LogoException;
import org.nlogo.nvm.Command;
import org.nlogo.nvm.EngineException;
import org.nlogo.nvm.Syntax;
import org.nlogo.window.GUIWorkspace;

public final strictfp class _load3Dshapes
	extends Command
{
	@Override
	public Syntax syntax()
	{
		int[] right = { Syntax.TYPE_STRING } ;
		return Syntax.commandSyntax( right , "O---" , true ) ;
	}
	@Override
	public void perform( final org.nlogo.nvm.Context context ) throws LogoException
	{
		String filename = argEvalString( context , 0 ) ;
		if( workspace instanceof GUIWorkspace )
		{
			try
			{
				( (GUIWorkspace) workspace ).addCustomShapes( filename ) ;
			}
			catch( java.io.IOException e )
			{
				throw new EngineException( context , this , e.getMessage() ) ;
			}
			catch( org.nlogo.shape.InvalidShapeDescriptionException e )
			{
				throw new EngineException( context , this , "Invalid shape file" ) ;
			}
		}
		context.ip = next ;
	}
}

File: src/main/org/nlogo/prim/gui/_load3Dshapes.scala
package org.nlogo.prim.gui

import org.nlogo.nvm.{ Command, Context, EngineException, Syntax }
import org.nlogo.shape.InvalidShapeDescriptionException
import org.nlogo.window.GUIWorkspace

class _load3Dshapes extends Command {
  override def syntax =
    Syntax.commandSyntax(Array(Syntax.TYPE_STRING), "O---", true)
  override def perform(context: Context) {
    val filename = argEvalString(context, 0)
    workspace match {
      case gw: GUIWorkspace =>
        try gw.addCustomShapes(filename)
        catch {
          case e: java.io.IOException =>
            throw new EngineException(context, this, e.getMessage)
          case e: InvalidShapeDescriptionException =>
            throw new EngineException(context, this, "Invalid shape file")
        }
    }
    context.ip = next
  }
}
